src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java

```diff
@@ -104,6 +113,54 @@ public Post updatePost(Long courseId, Post post) {
         return updatedPost;
     }
 
+    /**
+     * Checks course, user and post validity,
+     * updates the pin state, persists the post,
+     * and ensures that sensitive information is filtered out
+     *
+     * @param courseId  id of the course the post belongs to
+     * @param postId    id of the post to change the pin state for
+     * @param pinState  new boolean value of the pinned flag for the given post
+     * @return updated post that was persisted
+     */
+    public Post updatePinState(Long courseId, Long postId, Boolean pinState) {
+        Post post = postRepository.findByIdElseThrow(postId);
+
+        // update pin state
+        post.setPinned(pinState);
+        // ensure that pin state is compatible with archive state -> either one of them can be true
+        if (post.isArchived() && post.isPinned()) {
+            // if after archive state update both, archived and pinned, would be true the archive flag is flipped
+            post.setArchived(false);
+        }
+
+        return updatePost(courseId, post);
+    }
+
+    /**
+     * Checks course, user and post validity,
+     * updates the archive state, persists the post,
+     * and ensures that sensitive information is filtered out
+     *
+     * @param courseId      id of the course the post belongs to
+     * @param postId        id of the post to change the archive state for
+     * @param archiveState  new boolean value of the archived flag for the given post
+     * @return updated post that was persisted
+     */
+    public Post updateArchiveState(Long courseId, Long postId, Boolean archiveState) {
```

Comment:
Maybe consider doing this check as part of `Post::setArchived`?

Reply:
are you referring to the `isAtLeastTeachingAssistantInCourse()` check? If so, I guess the answer is given in [this comment](https://github.com/ls1intum/Artemis/pull/3907#discussion_r694881996) as well. If not, I do not understand the question 😄

Reply:
Mostly to the logic to update pin/archive state accordingly. Otherwise, maybe an enum type `NORMAL`, `PIN`, `ARCHIVE` would capture the state better than two booleans? However we can also save this for a later refactoring at this stage.

Reply:
done